include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -554,26 +596,48 @@ struct __usm_host_or_buffer_storage
     }
 
     auto
-    __get_acc(sycl::handler& __cgh)
+    __get_result_acc(sycl::handler& __cgh)
     {
-#if _ONEDPL_SYCL_UNIFIED_USM_BUFFER_PRESENT
-        return __usm ? __usm_host_or_buffer_accessor<_T>(__cgh, __usm_buf.get())
-                     : __usm_host_or_buffer_accessor<_T>(__cgh, __sycl_buf.get());
-#else
-        return sycl::accessor(*__sycl_buf.get(), __cgh, sycl::read_write, __dpl_sycl::__no_init{});
-#endif
+        if (__use_USM_host && __supports_USM_device)
+            return __usm_or_buffer_accessor<_T>(__cgh, __result_buf.get());
+        else if (__supports_USM_device)
+            return __usm_or_buffer_accessor<_T>(__cgh, __scratch_buf.get(), __scratch_n);
+        return __usm_or_buffer_accessor<_T>(__cgh, __sycl_buf.get(), __scratch_n);
     }
 
-    _T
-    __get_value(size_t idx = 0) const
+    auto
+    __get_scratch_acc(sycl::handler& __cgh)
     {
-        return __usm ? *(__usm_buf.get() + idx) : __sycl_buf->get_host_access(sycl::read_only)[idx];
+        if (__use_USM_host || __supports_USM_device)
+            return __usm_or_buffer_accessor<_T>(__cgh, __scratch_buf.get());
+        return __usm_or_buffer_accessor<_T>(__cgh, __sycl_buf.get());
     }
 
     bool
-    __get_usm() const
+    is_USM() const
     {
-        return __usm;
+        return __supports_USM_device;
+    }
+
+    // Note: this member function assumes the result is *ready*, since the __future has already
+    // waited on the relevant event.
+    _T
+    __get_value(size_t idx = 0) const
+    {
+        if (__use_USM_host && __supports_USM_device)
+        {
+            return *(__result_buf.get() + idx);
+        }
+        else if (__supports_USM_device)
+        {
+            _T __tmp;
+            __exec.queue().memcpy(&__tmp, __scratch_buf.get() + __scratch_n + idx, 1 * sizeof(_T)).wait();
```

Comment:
`wait` or `wait_and_throw` ?

Reply:
I think `wait` is more appropriate. Unless you feel that this is the right place to handle any async errors that could have occurred in the kernel. The future class uses wait and throw, this makes sense here as this is likely the last wait so you would want to collect any unconsumed errors produced. So I think it is fine as is.